### ISSUE TITLE
SAM records are output out of order

### ISSUE THREAD

**Issue by marcelm** [ksahlin/StrobeAlign#121]
It seems that the alignment records are not necessarily output in the order in which the reads appear in the input FASTQ. Judging from the output, it appears that a chunk that has been aligned by one of the worker threads is immediately sent to stdout.

This is a problem because two runs of StrobeAlign can lead to different outputs.

**Comment by ksahlin** [comment]
I see. I guess it is better to have deterministic output for testing. Is there a way to do this with minimal time overhead (for synchronising batches). I spent quite some time implementing the parallelisation, although one can probably improve it.

**Comment by marcelm** [comment]
Here is the code for this exact problem in Cutadapt: https://github.com/marcelm/cutadapt/blob/87a0d4025cdc9e2cc07256e300fcc62c49143b8f/src/cutadapt/pipeline.py#L882
I don’t know how well this applies here, but this is the idea:
- Input chunks are numbered consecutively.
- When a worker sends back its processed chunk, it also sends the chunk number.
- When the main thread receives a processed chunk for writing, it puts it into a dict.
- While the next chunk in sequence is available, it is written out.

There is very little runtime overhead with this method; it only needs some extra memory to hold the chunks that arrive out of order.

**Comment by ksahlin** [comment]
Yes, that sounds like a good solution with low time overhead! For larger enough batches, approximate worst case is that N batches are stored at the same time with N threads (most likely to happen at start). One can always tweak the batch size for performance. Currently it is at 100,000 I believe, which is quite large. I don't know the optimal size but with the solution above, probably smaller.

**Comment by marcelm** [comment]
Regarding the size of 100,000, I also think it is quite large. When StrobeAlign finishes processing a dataset, it works with fewer threads than the maximum for a relatively long time because there’s no new data to feed some of the threads. I would guess that the per-chunk overhead is quite small, so reducing this to 10,000 should be fine. But that needs some measurements.
